File: shared/src/commonMain/kotlin/com/skydown/shared/usecase/MerchandiseVariantResolver.kt
package com.skydown.shared.usecase

import com.skydown.shared.model.MerchandiseItem
import com.skydown.shared.model.MerchandiseVariant
import com.skydown.shared.text.SharedText

object MerchandiseVariantResolver {
    fun availableSizes(item: MerchandiseItem): List<String> {
        return item.variants
            .mapNotNull { it.size?.trim()?.takeIf(String::isNotEmpty) }
            .distinctBy { it.lowercase() }
    }

    fun availableColors(item: MerchandiseItem, size: String?): List<String> {
        val normalizedSize = size?.normalizedValue()
        return item.variants
            .filter { variant ->
                normalizedSize == null || variant.size.normalizedValue() == normalizedSize
            }
            .mapNotNull { it.color?.trim()?.takeIf(String::isNotEmpty) }
            .distinctBy { it.lowercase() }
    }

    fun resolveVariant(
        item: MerchandiseItem,
        size: String,
        color: String?,
    ): Result<MerchandiseVariant> {
        if (item.variants.isEmpty()) {
            return Result.failure(
                IllegalStateException(SharedText.MERCH_VARIANT_NONE_AVAILABLE.format(item.name)),
            )
        }

        val normalizedSize = size.normalizedValue()
            ?: return Result.failure(IllegalArgumentException(SharedText.MERCH_VARIANT_SIZE_REQUIRED))
        val normalizedColor = color.normalizedValue()

        val matches = item.variants.filter { variant ->
            variant.size.normalizedValue() == normalizedSize &&
                variant.color.normalizedValue() == normalizedColor
        }

        if (matches.size == 1) {
            return Result.success(matches.first())
        }

        if (matches.isEmpty()) {
            val colorPart = color?.let { " / $it" }.orEmpty()
            return Result.failure(
                IllegalArgumentException(SharedText.MERCH_VARIANT_NOT_FOUND.format(size, colorPart)),
            )
        }

        val colorPart = color?.let { " / $it" }.orEmpty()
        return Result.failure(
            IllegalStateException(SharedText.MERCH_VARIANT_MULTIPLE_FOUND.format(size, colorPart)),
        )
    }
}

private fun String?.normalizedValue(): String? {
    return this
        ?.trim()
        ?.takeIf { it.isNotEmpty() }
        ?.lowercase()
}

File: Skydown App/Services/Merchandise/MerchandiseVariantResolver.swift
import Foundation

enum MerchandiseVariantResolver {
    static func availableSizes(for item: MerchandiseItem) -> [String] {
        deduplicated(
            item.variants.compactMap { $0.size?.trimmedNonEmpty }
        )
    }

    static func availableColors(for item: MerchandiseItem, size: String?) -> [String] {
        let normalizedSize = size?.normalizedVariantValue
        return deduplicated(
            item.variants.compactMap { variant in
                guard normalizedSize == nil || variant.size?.normalizedVariantValue == normalizedSize else {
                    return nil
                }
                return variant.color?.trimmedNonEmpty
            }
        )
    }

    static func resolveVariant(
        for item: MerchandiseItem,
        size: String,
        color: String?
    ) throws -> MerchandiseVariant {
        guard !item.variants.isEmpty else {
            throw NSError(
                domain: "MerchandiseVariantResolver",
                code: 404,
                userInfo: [NSLocalizedDescriptionKey: "Für \(item.name) sind keine Shopify-Varianten hinterlegt."]
            )
        }

        guard let normalizedSize = size.normalizedVariantValue else {
            throw NSError(
                domain: "MerchandiseVariantResolver",
                code: 400,
                userInfo: [NSLocalizedDescriptionKey: "Bitte wähle eine Größe."]
            )
        }

        let normalizedColor = color?.normalizedVariantValue
        let matches = item.variants.filter { variant in
            variant.size?.normalizedVariantValue == normalizedSize &&
            variant.color?.normalizedVariantValue == normalizedColor
        }

        if let exactMatch = matches.only {
            return exactMatch
        }

        if matches.isEmpty {
            let detail = color?.trimmedNonEmpty.map { " / \($0)" } ?? ""
            throw NSError(
                domain: "MerchandiseVariantResolver",
                code: 404,
                userInfo: [NSLocalizedDescriptionKey: "Keine passende Variante für \(size)\(detail) gefunden."]
            )
        }

        throw NSError(
            domain: "MerchandiseVariantResolver",
            code: 409,
            userInfo: [NSLocalizedDescriptionKey: "Mehrere Varianten für diese Auswahl gefunden."]
        )
    }

    private static func deduplicated(_ values: [String]) -> [String] {
        var seen: Set<String> = []
        return values.filter {
            let normalized = $0.lowercased()
            return seen.insert(normalized).inserted
        }
    }
}

private extension String {
    var trimmedNonEmpty: String? {
        let trimmed = trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    var normalizedVariantValue: String? {
        trimmedNonEmpty?.lowercased()
    }
}

private extension Array {
    var only: Element? {
        count == 1 ? first : nil
    }
}
